### PR TITLE
Direct GI

### PR DESCRIPTION
**Pull request by VZout**
Added direct GI from before the vacation.

### CODE REVIEW

File: resources/shaders/path_tracer.hlsl [hunk 1]
```diff
@@ -139,6 +139,66 @@ float probabilityToSampleDiffuse(float3 difColor, float3 specColor)
...
+	float3 F = F_Schlick(LdotH, metal, albedo);
+	//float3 F = F_ShlickSimple(metal, LdotH);
+
+	float3 kS = F_SchlickRoughness(NdotV, metal, albedo, roughness);
+	float3 kD = (1.f - kS) * (1.0 - metal);
+	float3 spec = (D * F * G) / (4.0 * NdotV * NdotL + 0.001f);
+
+	//prefiltered_color * (kS * sampled_brdf.x + sampled_brdf.y);
+	return shadow_mult * (light_intensity * (NdotL * spec + NdotL * albedo / M_PI));
+	return shadow_mult * (kD * albedo / PI + spec) * light_intensity * NdotL;
```
Comment (Nielsbishere) [inline resources/shaders/path_tracer.hlsl]: Why is there two returns here?

File: resources/shaders/path_tracer.hlsl [hunk 2]
```diff
@@ -96,7 +96,7 @@ float4 TraceColorRay(float3 origin, float3 direction, unsigned int depth, unsign
 	RayDesc ray;
 	ray.Origin = origin;
 	ray.Direction = direction;
-	ray.TMin = EPSILON;
+	ray.TMin = 0;
```
Comment (Nielsbishere) [inline resources/shaders/path_tracer.hlsl]: Won't this cause self-intersection?

File: resources/shaders/accumulation.hlsl [hunk 3]
```diff
@@ -17,11 +17,7 @@ void main(uint3 DTid : SV_DispatchThreadID)
 	float2 resolution;
 	input.GetDimensions(resolution.x, resolution.y);
 	
-	float2 uv = float2(DTid.xy) / resolution;
-	float gamma = 2.2;
-	float exposure = 1;
-
-	float3 color = input[DTid.xy].rgb / (frame_idx);
+	float3 color = input[DTid.xy].rgb / (frame_idx + 1);
```
Comment (Nielsbishere) [inline resources/shaders/accumulation.hlsl]: Why do you divide by frame_idx? Means the color would be divided by 1, then 2, then 3?

File: resources/shaders/path_tracer.hlsl [hunk 4]
```diff
@@ -139,6 +139,66 @@ float probabilityToSampleDiffuse(float3 difColor, float3 specColor)
 	return lumDiffuse / (lumDiffuse + lumSpecular);
 }
 
+float3 ggxDirect(float3 hit_pos, float3 fN, float3 N, float3 V, float3 albedo, float metal, float roughness, unsigned int seed, unsigned int depth)
+{
+	// #################### GGX #####################
+	uint light_count = lights[0].tid >> 2;
+	int light_to_sample = min(int(nextRand(seed) * light_count), light_count - 1);
+	Light light = lights[light_to_sample];
```
Comment (Nielsbishere) [inline resources/shaders/path_tracer.hlsl]: What if there's no lights available?

File: resources/shaders/path_tracer.hlsl (hunk 2 above)
Comment (Nielsbishere) [reply resources/shaders/path_tracer.hlsl]: Nvm, see you added it to the origin

File: resources/shaders/accumulation.hlsl (hunk 3 above)
Comment (VZout) [reply resources/shaders/accumulation.hlsl]: Yes

File: resources/shaders/path_tracer.hlsl (hunk 4 above)
Comment (VZout) [reply resources/shaders/path_tracer.hlsl]: Good point. Forgot about that case.

File: resources/shaders/path_tracer.hlsl (hunk 1 above)
Comment (VZout) [reply resources/shaders/path_tracer.hlsl]: removed

File: resources/shaders/accumulation.hlsl (hunk 3 above)
Comment (Nielsbishere) [reply resources/shaders/accumulation.hlsl]: Hmm, I guess the var name is correct, but frame_idx is also referenced to as the current render index

**Pull request merged**